First draft of metrics feature

BMorinBlitz opened the pull request:
First draft of a metrics feature for Riven.

Because the external API doesn't have an easy way to capture the operation string and route without manually adding code to every call point, I'm proposing an optional metrics feature that creates a histogram for every Riot API call.  This makes it easier to root cause issues talking to the Riot API.

`metrics.rs` wasn't an easy write for me, might be a better way to express this.  Biased for something simple that minimized risk of something internal leaking to the public API.

I'm also unsure if there is a better way to split out this feature in `Cargo.toml`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

BMorinBlitz committed:
First draft of metrics feature

MingweiSamuel commented on riven/src/metrics.rs:
```diff
@@ -0,0 +1,57 @@
+}
+
+/// Returns a wrapped future that records the time it takes to complete the future as a histogram metric.
+pub fn timed<Fut>(
+    future: Fut,
+    operation: &'static str,
+    route: &'static str,
+) -> impl Future<Output = Fut::Output>
+where
+    Fut: Future,
+{
+    TimedMetric {
+        inner: future,
+        operation,
+        route,
+        start: None,
+    }
+}
```
I suspect we can simplify by using an async method instead of manually writing the future, avoids needing the pin-project crate
```rust
async {
    let start = Instant::now();
    let out = future.await;
    metrics::histogram!("riot_api", "operation" => operation, "route" => route)
        .record(start.elapsed());
    out
}
```

MingweiSamuel commented on riven/src/metrics.rs:
```diff
@@ -0,0 +1,57 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Instant;
```
`crate::time::Instant` for wasm compatibility (assuming metrics works on wasm)

MingweiSamuel reviewed (commented):
Appreciate the PR!

BMorinBlitz committed:
Use an async block per PR feedback

BMorinBlitz commented:
Updated PR to use an async block per feedback.  That removed the `pin-project` dependency and simplified `metrics.rs`.

Also pointed `Instant` to `crate::time::Instant` per feedback.

bmorin commented:
@MingweiSamuel Anything else I can do or feedback I can address?

BMorinBlitz commented:
Any additional feedback or anything I can do to help?

BMorinBlitz commented:
Thank You!
